File: tensorflow/python/platform/stacktrace_handler_test.py
# Copyright 2018 The TensorFlow Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# =============================================================================
"""Test to make sure stack trace is generated in case of test failures."""

from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import argparse
import os
import signal
import subprocess
import sys

from tensorflow.python.platform import test
from tensorflow.python.platform import tf_logging as logging


# FLAGS defined at the bottom:
# child (bool) set to true if we are running in the child process.
FLAGS = None

_CHILD_FLAG_HELP = 'Boolean. Set to true if this is the child process.'


class StacktraceHandlerTest(test.TestCase):

  def testChildProcessKillsItself(self):
    if FLAGS.child:
      os.kill(os.getpid(), signal.SIGABRT)

  def testGeneratesStacktrace(self):
    if FLAGS.child:
      return

    # Subprocess sys.argv[0] with --child=True
    if sys.executable:
      child_process = subprocess.Popen(
          [sys.executable, sys.argv[0], '--child=True'], cwd=os.getcwd(),
          stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    else:
      child_process = subprocess.Popen(
          [sys.argv[0], '--child=True'], cwd=os.getcwd(),
          stdout=subprocess.PIPE, stderr=subprocess.PIPE)

    # Capture its output. capture both stdout and stderr and append them.
    # We are not worried about timing or order of messages in this test.
    child_stdout, child_stderr = child_process.communicate()
    child_output = child_stdout + child_stderr

    # Make sure the child process is dead before we proceed.
    child_process.wait()

    logging.info('Output from the child process:')
    logging.info(child_output)

    # Verify a stack trace is printed.
    self.assertIn(b'PyEval_EvalFrame', child_output)


if __name__ == '__main__':
  parser = argparse.ArgumentParser()
  parser.add_argument(
      '--child', type=bool, default=False, help=_CHILD_FLAG_HELP)
  FLAGS, unparsed = parser.parse_known_args()

  # Now update argv, so that unittest library does not get confused.
  sys.argv = [sys.argv[0]] + unparsed
  test.main()

File: tensorflow/core/platform/stacktrace_handler_test.cc
/* Copyright 2015 The TensorFlow Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/
// Testing proper operation of the stacktrace handler.

#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <unistd.h>
#include <string>

#include "tensorflow/core/platform/logging.h"
#include "tensorflow/core/platform/test.h"

namespace tensorflow {
namespace {

#define READ_BUFFER_SIZE 1024

TEST(StacktraceHandlerTest, GeneratesStacktrace) {
  // Create a pipe to write/read the child stdout.
  int test_pipe[2];
  EXPECT_EQ(pipe(test_pipe), 0);

  // Fork the process.
  int test_pid = fork();

  if (test_pid == 0) {
    // Child process.
    // Close the read end of the pipe, redirect stdout and sleep.
    close(test_pipe[0]);
    dup2(test_pipe[1], STDOUT_FILENO);
    dup2(test_pipe[1], STDERR_FILENO);
    sleep(10);
  } else {
    // Parent process.
    // Close the write end of the pipe, wait a little and send SIGABRT to the
    // child process. Then watch the pipe.
    close(test_pipe[1]);
    sleep(1);

    // Send the signal.
    kill(test_pid, SIGABRT);

    // Read from the pipe.
    char buffer[READ_BUFFER_SIZE];
    std::string child_output = "";
    while (true) {
      int read_length = read(test_pipe[0], buffer, READ_BUFFER_SIZE);
      if (read_length > 0) {
        child_output += std::string(buffer, read_length);
      } else {
        break;
      }
    }
    close(test_pipe[0]);

    // Just make sure we can detect one of the calls in testing stack.
    string test_stack_frame = "testing::internal::UnitTestImpl::RunAllTests()";

    // Print the stack trace detected for information.
    LOG(INFO) << "Output from the child process:";
    LOG(INFO) << child_output;

    EXPECT_NE(child_output.find(test_stack_frame), std::string::npos);
  }
}

}  // namespace
}  // namespace tensorflow
